### DISCUSSION TITLE
[Reading Group]: Learning Transformer Programs

### DISCUSSION THREAD

**Discussion by nicaudinet** [dsai-nlp/core#23] [Meetings & Announcements]
### 📆 Date & Time

15:00-17:00 on Tuesday 23rd of September 2025

### 📍 Location or Link

8103

### 📄 Paper Link

https://proceedings.neurips.cc/paper_files/paper/2023/hash/995f693b73050f90977ed2828202645c-Abstract-Conference.html

### 📖 Curator

Nicolas Audinet de Pieuchon

### 🗒️ Agenda

# Paper Summary

This paper introduces a modified version of the transformer architecture which can be automatically converted to a program in the RASP programming language. They term these "transformer programs". In the paper they learn transformer programs for a variety of tasks and argue that they are mechanistically interpretable by design. They even convert the transformer programs to Python and use code analysis tools to debug model errors and find circuits.

# Agenda

Discuss the questions, concerns, and potential follow-up opportunities raised in the comments!

**Comment by EkaterinaVoloshina** [comment]
I really liked the idea of the paper! I like that it produces an easy-to-understand and reproducible result (i.e. one can ideally use learned programs instead of Transformer models).

*Questions*: on page 9 it says the model performed on-par but it surpassed both models in Table 2 or did I miss something? 

*Concerns*: 
1. would that work on MoE, at least to some extent? 
2. Hard attention seems more of the theoretical concept rather than usable technique (they discuss similar issues at the end though)

**Reply by ricj** [reply]
Re Table 2: yes it is a bit weird that they say that it is "on par" with the standard Transformer when it is in fact somewhat better.

On the other hand this seems like an exceptionally poor performance for NER with the CoNLL dataset (where one would typically expect an F-score above 0.9) so I'm not sure what they do with their standard Transformer here...

**Reply by ricj** [reply]
Re 2: I also share the impression that hard attention has mostly been of interest to the various lines of research trying to relate Transformers to formal languages e.g. https://papers.nips.cc/paper_files/paper/2024/file/13d7f172259b11b230cc5da8768abc5f-Paper-Conference.pdf

**Reply by ricj** [reply]
Re 1: my hunch would be that MoE layers can be expressed in their framework. I think it would just make the MLP part a bit more complicated. In fact MoE implementations such as the Switch Transformer (https://arxiv.org/pdf/2101.03961) involve a discrete choice of expert (1 expert is active in each MoE layer) that's a bit similar to the hard attention.

Currently, their feedforward computation is 

$z = \text{One-hot}(\arg\max(f_θ(xW_{in}))$

so I suppose a single-expert MoE layer would be something like

$e = \arg\max(x R))$
$z = \text{One-hot}(\arg\max(f_θ(xW_{e}))$

**Comment by ricj** [comment]
Why do they say in §4.1 that it is "in-context learning"? It seems like a regular supervised task?

**Comment by nicaudinet** [comment]
Questions:
- They formalise `select` with different $n$ and $m$, but in the original paper $n = m$. In what cases does $n \neq m$ here?
- What is Gumbel reparametrisation?
- In appendix C.3 they note that it seems that attention layers are more difficult to optimise, and they hypothesise that they might get stuck in local minima more easily. Why would this be the case?

Comments and future work:
- Can we use these methods to track how a transformer changes during training?
- What other interesting modules could we add?
- Would this approach benefit from programming languages like Haskell where one has access to equational reasoning?
- As well as making transformers more understandable to humans, this seems to make transformers more understandable to LLMs as well
- Can we relax some of their constraints to make this method more scalable?

**Reply by ricj** [reply]
1) Something to do with self/cross-attention?

2) The Gumbel softmax is a trick that allows one to backprop in a computational graph that involves sampling from a categorical distribution https://medium.com/data-science/what-is-gumbel-softmax-7f6d9cdcb90e
